### PR TITLE
 Add initial implementation for the AWS Marketplace Metering connector

### PR DESCRIPTION
**Pull request by ayeshLK**
Part of: [#6865](https://github.com/ballerina-platform/ballerina-library/issues/6865)

### CODE REVIEW

File: native/src/main/java/io/ballerina/lib/aws/mpm/Constants.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * under the License.
+ */
+
+package io.ballerina.lib.aws.mpm;
+
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BString;
+
+/**
+ * Represents the constants related to Ballerina MPE connector.
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/aws/mpm/Constants.java]:
```suggestion
 * Represents the constants related to Ballerina MPM connector.
```

File: native/src/main/java/io/ballerina/lib/aws/mpm/Constants.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    BString MPM_BATCH_METER_USAGE_RESPONSE_RESULTS = StringUtils.fromString("results");
+    BString MPM_BATCH_METER_USAGE_RESPONSE_UNPROC_RECORDS = StringUtils.fromString("unprocessedRecords");
+
+    // Constants related to MPM `UsageRecordResult`
+    String MPM_USAGE_RECORD_RESULT = "UsageRecordResult";
+    BString MPM_USAGE_RECORD_RESULT_METERING_RECORD = StringUtils.fromString("meteringRecordId");
+    BString MPM_USAGE_RECORD_RESULT_STATUS = StringUtils.fromString("status");
+    BString MPM_USAGE_RECORD_RESULT_USAGE_RECORD = StringUtils.fromString("usageRecord");
+
+    // Constants related to MPE Error
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/aws/mpm/Constants.java]:
```suggestion
    // Constants related to MPM Error
```

File: ballerina/client.bal [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        return self.externBatchMeterUsage(validated);
+    }
+
+    isolated function externBatchMeterUsage(BatchMeterUsageRequest request) returns BatchMeterUsageResponse|Error = 
+    @java:Method {
+        name: "batchMeterUsage",
+        'class: "io.ballerina.lib.aws.mpm.AwsMpmClient"
+    } external;
+
+    # Closes the AWS MPE client resources.
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal]:
```suggestion
    # Closes the AWS MPM client resources.
```

File: native/src/main/java/io/ballerina/lib/aws/mpm/AwsMpmClient.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    private static final ExecutorService EXECUTOR_SERVICE = Executors.newCachedThreadPool(new AwsMpmThreadFactory());
+
+    private AwsMpmClient() {
+    }
+
+    /**
+     * Creates an AWS MPM native client with the provided configurations.
+     *
+     * @param bAwsMpmClient The Ballerina AWS MPM client object.
+     * @param configurations AWS MPE client connection configurations.
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/aws/mpm/AwsMpmClient.java]:
```suggestion
     * @param configurations AWS MPM client connection configurations.
```

**Review by NipunaRanasinghe — DISMISSED** [review]
LGTM

File: native/src/main/java/io/ballerina/lib/aws/mpm/AwsMpmClient.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+import software.amazon.awssdk.services.marketplacemetering.model.BatchMeterUsageRequest;
+import software.amazon.awssdk.services.marketplacemetering.model.BatchMeterUsageResponse;
+import software.amazon.awssdk.services.marketplacemetering.model.ResolveCustomerRequest;
+import software.amazon.awssdk.services.marketplacemetering.model.ResolveCustomerResponse;
+
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public final class AwsMpmClient {
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/aws/mpm/AwsMpmClient.java]: Same as https://github.com/ballerina-platform/module-ballerinax-aws.marketplace.mpe/pull/2#discussion_r1722770844. Thoughts?
```suggestion
public final class Client {
```

Comment (ayeshLK) [reply native/src/main/java/io/ballerina/lib/aws/mpm/AwsMpmClient.java]: Fixed with [aea2f01eb6ce24ba0ccf85cb04b84993b28b9f79](https://github.com/ballerina-platform/module-ballerinax-aws.marketplace.mpm/pull/2/commits/aea2f01eb6ce24ba0ccf85cb04b84993b28b9f79)

**Review by NipunaRanasinghe — DISMISSED** [review]
LGTM

**Review by NipunaRanasinghe — DISMISSED** [review]
LGTM

**Review by Nuvindu — APPROVED** [review]

**Pull request merged**